### ISSUE TITLE
NullPointerException encountered with Spring Cloud Stream Binder for Kinesis when enabling observation and using StreamBridge

### ISSUE THREAD

**Issue by kabennett** [spring-cloud/spring-cloud-stream#2889]
We are encountering a NullPointerException with the use of StreamBridge to send a message, observation enabled, and the use of Spring Cloud Stream Binder for Kinesis version 4.0.0. To get a background on this issue, please see [this stackoverflow](https://stackoverflow.com/questions/75102667/propagating-trace-ids-across-http-rest-and-kinesis-messaging-service-boundaries). Please note Artem Bilan's [comment ](https://stackoverflow.com/questions/75102667/propagating-trace-ids-across-http-rest-and-kinesis-messaging-service-boundaries#comment137275390_75114280) at the end of the end of the stackoverflow post stating that this issue should be mentioned here.

1. Use a ```StreamBridge``` to send a message to Kinesis.
2. Ensure spring.cloud.stream.kinesis.binder.enable-observation=true
3. Without the ```NewDestinationBindingCallback``` bean defined, which I will provide below, a NullPointerException is encountered as seen in the attached screen shot.

**Version of the framework**
Spring Boot 3.2.2
Spring Cloud 2023.0.0
Spring Cloud Stream Binder for Kinesis 4.0.0

**Expected behavior**
No exceptions thrown and Trace ID should be propagated from producer to consumer.

**Screenshots**
<img width="647" alt="stracktrace" src="https://github.com/spring-cloud/spring-cloud-stream/assets/15836610/c4d5eb72-f783-4457-ab32-5c888f1bbc27">

**Additional context**
This is the bean I add to my microservice to get everything working as expected:

```
@Bean
public NewDestinationBindingCallback newDestinationBindingCallback(
    ObservationRegistry observationRegistry) {
  return (channelName, channel, producerProperties, extendedProducerProperties) -> {
    var abstractMessageChannel = ((AbstractMessageChannel) channel);
    abstractMessageChannel.setBeanName(channelName);
    abstractMessageChannel.registerObservationRegistry(observationRegistry);
  };
}
```

So, I guess the question to ask is whether this bean can/should be added to Spring Cloud Stream conditionally so that anyone trying to enable observation and use a ```StreamBridge``` to send a message over Kinesis doesn't encounter the same NPE?

**Comment by sobychacko** [comment]
@kabennett This sounds like a reasonable thing to do. What should the `@Conditional` be based on? `ObservationRegistry`?

**Comment by kabennett** [comment]
Yes, I would think it makes sense to base it on ```ObservationRegistry```. One thing I haven't mentioned is that we **_have not_** run into this problem when using the Kafka binder. In fact, we have both Kafka and Kinesis streams at this point in our application architecture (i.e., we can consume/produce from either Kafka or Kinesis arbitrarily with the multi-binder features of Spring Cloud Stream), and this issue only occurs when using ```StreamBridge``` to send a message to Kinesis, never to Kafka, and I've tested with both. So, I was surprised by Artem's comment that this is out of scope for the Kinesis binder and that this bean would belong in Spring Cloud Stream core.

**Comment by artembilan** [comment]
Probably Kafka Binder does not have an observation handling. 
Otherwise it would be great to have some simple project to reproduce. And yes, with a confirmation that Kafka side is indeed with an observation enabled. Otherwise we don’t compare apples with apples.
Thanks

**Comment by artembilan** [comment]
The `StreamBridge` has a logic like this in its `send()`:
```
			else {
				messageChannel = new DirectWithAttributesChannel();
...
				this.bindingService.bindProducer(messageChannel, destinationName, true, binder);
			}
```

The Kinesis Binder does this then in its `bindProducer()` logic:
```
((IntegrationManagement) outputChannel).registerObservationRegistry(this.observationRegistry);
```

The Kafka Binder does that like this instead:
```
kafkaTemplate.setObservationEnabled(this.configurationProperties.isEnableObservation());
```
Where, even if it is created like this:
```
KafkaTemplate<byte[], byte[]> kafkaTemplate = new KafkaTemplate<>(producerFB);
```
it has some default:
```
private String beanName = "kafkaTemplate";
```
That's why it does not fail.

However this does not feel right, since all the out bindings are going to have same `spring.kafka.template.name` observation tag.

I think the fix is supposed to be like this in the `StreamBridge` by itself:
```
DirectWithAttributesChannel directChannel = new DirectWithAttributesChannel();
directChannel.setBeanName(destinationName);
```
Kinda exactly exactly what you did with the `NewDestinationBindingCallback`, but now you would not need to do that since the framework would cover you.

As as separate issue I would revise that `KafkaTemplate` stuff in the Kafka Binder, since that in my opinion has to be fixed as well.

Thanks

**Comment by sobychacko** [comment]
Artem, that sounds like close to a change I made last week. https://github.com/spring-cloud/spring-cloud-stream/commit/c72c5487631a65248baa44d1b5f4afdf53c25854

**Comment by artembilan** [comment]
Yep! That's the one! 

Closed as `Duplicated` of #2885
